drivers/sensor/xbr818/xbr818.c

```diff
@@ -0,0 +1,533 @@
+		if (ret != 0) {
+			break;
+		}
+		val->val1 |= tmp << 8;
+		ret = i2c_reg_read_byte_dt(&config->i2c, XBR818_DELAY_TIME_3, &tmp);
+		if (ret != 0) {
+			break;
+		}
+		val->val1 |= tmp << 16;
+		tmpf = (uint64_t)val->val1 * 1000000;
+		tmpf /= SENSOR_XBR818_CLOCKRATE;
+		val->val1 = tmpf / 1000000;
+		val->val2 = tmpf - val->val1 * 1000000;
+		break;
+	case SENSOR_ATTR_XBR818_LOCK_TIME:
+		ret = i2c_reg_read_byte_dt(&config->i2c, XBR818_LOCK_TIME_1, &tmp);
+		if (ret != 0) {
+			break;
+		}
+		val->val1 = tmp & 0xFF;
+		ret = i2c_reg_read_byte_dt(&config->i2c, XBR818_LOCK_TIME_2, &tmp);
+		if (ret != 0) {
+			break;
+		}
+		val->val1 |= tmp << 8;
+		ret = i2c_reg_read_byte_dt(&config->i2c, XBR818_LOCK_TIME_3, &tmp);
+		if (ret != 0) {
+			break;
+		}
+		val->val1 |= tmp << 16;
+		tmpf = (uint64_t)val->val1 * 1000000;
+		tmpf /= SENSOR_XBR818_CLOCKRATE;
+		val->val1 = tmpf / 1000000;
+		val->val2 = tmpf - val->val1 * 1000000;
+		break;
+	case SENSOR_ATTR_XBR818_RF_POWER:
+		ret = i2c_reg_read_byte_dt(&config->i2c, XBR818_RF_POWER, &tmp);
+		if (ret != 0) {
+			break;
+		}
+		val->val1 = tmp & 0x7;
+		break;
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+		ret = i2c_reg_read_byte_dt(&config->i2c, XBR818_SAMPLE_RATE_DIVIDER, &tmp);
+		if (ret != 0) {
+			break;
+		}
+		val->val1 = SENSOR_XBR818_CLOCKRATE / tmp;
+		break;
+	default:
+		ret = -ENODEV;
+
+		break;
+	}
+
+	ret = xbr818_disable_i2c(dev);
+	if (ret != 0) {
+		return ret;
+	}
+
+	return ret;
+}
+
+static void xbr818_work(struct k_work *work)
+{
+	struct xbr818_data *data = CONTAINER_OF(work, struct xbr818_data, work);
+
+	if (likely(data->handler != NULL)) {
+		data->handler(data->dev, data->trigger);
+	}
+}
+
+static void xbr818_gpio_callback(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	struct xbr818_data *data = CONTAINER_OF(cb, struct xbr818_data, gpio_cb);
+
+	k_work_submit(&data->work);
+}
+
+static int xbr818_trigger_set(const struct device *dev, const struct sensor_trigger *trig,
+			      sensor_trigger_handler_t handler)
+{
+	const struct xbr818_config *config = dev->config;
+	struct xbr818_data *data = dev->data;
+	int ret;
+
+	if (trig->chan != SENSOR_CHAN_PROX) {
+		LOG_ERR("%s: requesting unsupported channel %i", dev->name, trig->chan);
+		return -ENOTSUP;
+	}
+
+	if (trig->type != SENSOR_TRIG_MOTION) {
+		LOG_ERR("%s: requesting unsupported trigger %i", dev->name, trig->type);
+		return -ENOTSUP;
+	}
+
+	data->handler = handler;
+	data->trigger = trig;
+	ret = gpio_pin_interrupt_configure_dt(&config->io_val, GPIO_INT_EDGE_RISING);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (handler) {
+		ret = gpio_add_callback(config->io_val.port, &data->gpio_cb);
+	} else {
+		ret = gpio_remove_callback(config->io_val.port, &data->gpio_cb);
+	}
+
+	return ret;
+}
+
+static int xbr818_init_defaults(const struct device *dev)
+{
+	const struct xbr818_config *config = dev->config;
+	int ret;
+
+	ret = i2c_reg_write_byte_dt(&config->i2c, XBR818_IO_ACTIVE_VALUE_REG, 0x03);
```

Comment:
nit: This is a lot of branches. Generally if one fails we can't recover anyway. I usually like to do:

```
int ret = 0;

ret |= i2c_*(...);
__ASSERT(ret == 0, "Some error message);

ret |= i2c_*(...);
__ASSERT(ret == 0, "Some error message);

return ret;
```

This way when I build a debug build I enable the asserts and I can see exactly when something fails, but in production I just care if the function failed and the asserts turn into no-ops.

Reply:
This is nice, done.